File: gantt-backend/app/api/push.py
import fastapi
from sqlalchemy import orm

from app.api import auth
from app.core import db, exception
from app.schemas import push as push_schemas
from app.services import push_service

router = fastapi.APIRouter(prefix="/api/push", tags=["Push Notifications"])


@router.post("/subscribe", response_model=push_schemas.PushSubscriptionResponse)
def subscribe_to_push(
    subscription_data: push_schemas.PushSubscriptionCreate,
    current_user=fastapi.Depends(auth.get_current_user),
    data_base: orm.Session = fastapi.Depends(db.get_db),
):
    """Подписаться на push уведомления"""
    subscription = push_service.create_push_subscription_service(
        data_base,
        user_id=current_user.id,
        endpoint=subscription_data.endpoint,
        p256dh=subscription_data.p256dh,
        auth=subscription_data.auth
    )
    return subscription


@router.get("/subscriptions", response_model=list[push_schemas.PushSubscriptionResponse])
def get_subscriptions(
    current_user=fastapi.Depends(auth.get_current_user),
    data_base: orm.Session = fastapi.Depends(db.get_db),
):
    """Получить все подписки пользователя"""
    return push_service.get_user_subscriptions_service(data_base, current_user.id)


@router.delete("/subscriptions/{subscription_id}", status_code=204)
def delete_subscription(
    subscription_id: int,
    current_user=fastapi.Depends(auth.get_current_user),
    data_base: orm.Session = fastapi.Depends(db.get_db),
):
    """Удалить подписку"""
    try:
        push_service.delete_push_subscription_service(
            data_base,
            subscription_id,
            current_user.id
        )
    except exception.NotFoundError as e:
        raise fastapi.HTTPException(404, str(e))
    except exception.ForbiddenError as e:
        raise fastapi.HTTPException(403, str(e))

File: gantt-frontend/src/api/push.js
import axios from "axios";

export async function subscribeToPushApi(subscription, token) {
  try {
    const keys = subscription.getKey
      ? {
          p256dh: btoa(
            String.fromCharCode(
              ...new Uint8Array(subscription.getKey("p256dh")),
            ),
          ),
          auth: btoa(
            String.fromCharCode(...new Uint8Array(subscription.getKey("auth"))),
          ),
        }
      : subscription.keys;

    const response = await axios.post(
      "/api/push/subscribe",
      {
        endpoint: subscription.endpoint,
        p256dh: keys.p256dh,
        auth: keys.auth,
      },
      { headers: { Authorization: token } },
    );
    return { ok: true, subscription: response.data };
  } catch (e) {
    return { ok: false, status: e.response?.status };
  }
}

export async function getPushSubscriptionsApi(token) {
  try {
    const response = await axios.get("/api/push/subscriptions", {
      headers: { Authorization: token },
    });
    return { ok: true, subscriptions: response.data };
  } catch (e) {
    return { ok: false, status: e.response?.status };
  }
}

export async function deletePushSubscriptionApi(subscriptionId, token) {
  try {
    await axios.delete(`/api/push/subscriptions/${subscriptionId}`, {
      headers: { Authorization: token },
    });
    return { ok: true };
  } catch (e) {
    return { ok: false, status: e.response?.status };
  }
}
